sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala

```diff
@@ -364,4 +377,82 @@ object HostAlloc {
       }
     }
   }
+
+
+  /**
+   * For bookkeeping host memory usage per thread
+   */
+  private trait PerThreadMemoryUsage {
+    def add(addr: Long, amount: Long, callstack: String): Unit
+    def remove(addr: Long, amount: Long): Unit
+  }
+  private class SimplePerThreadMemoryUsage extends PerThreadMemoryUsage {
+    private val totalMem: LongAdder = new LongAdder()
+    override def toString: String = s"${totalMem.sum()} bytes in total"
+    override def add(addr: Long, amount: Long, callstack: String): Unit = {
+      totalMem.add(amount)
+    }
+    override def remove(addr: Long, amount: Long): Unit = totalMem.add(-amount)
+  }
+  private case class MemoryUsageDetail(addr: Long, amount: Long, callStack: String) {
+    override def toString: String = s"$amount bytes behind address $addr at $callStack"
+  }
+
+  private class PerThreadMemoryUsageInDetails extends PerThreadMemoryUsage {
+    private val details: mutable.Map[Long, MemoryUsageDetail] = mutable.Map()
+    override def toString: String =
+      s"Total ${details.values.map(_.amount).sum} bytes from below callstacks:\n" +
+        s"${details.values.mkString("\n")}"
+
+    override def add(addr: Long, amount: Long, callstack: String): Unit =
+      details.put(addr, MemoryUsageDetail(addr, amount, callstack))
+
+    override def remove(addr: Long, amount: Long): Unit =
+      details.remove(addr)
+  }
+  private val muPerThreads = new ConcurrentHashMap[Long, PerThreadMemoryUsage]()
+  private val addr2threadId = new ConcurrentHashMap[Long, java.lang.Long]()
+
+  private def bookkeepHostMemoryAlloc(addr: Long, threadId: Long, amount: Long): Unit = {
+    HostAlloc.addr2threadId.put(addr, threadId)
```

Comment:
nit: It is rather confusing that for bookkeepHostMemoryAlloc the check for BOOKKEEP_MEMORY is the responsibility of the caller, but for bookkeepHostMemoryFree it is not.  Could we make this consistent?

Reply:
fixed